### PR TITLE
Kfold + dataset update

### PR DESCRIPTION
**Pull request by justusschock**
Add methods to extract subsets from dataset and datamanager and update kfold accordingly.

### CODE REVIEW

File: delira/training/experiment.py [hunk 1]
```diff
@@ -94,17 +94,31 @@ class initialization
...
+        The two supported scenarios are:
+
+            * passing a list of datamanagers: the list will be splitted and 
+            multiple datamanagers will be combinded into a single one 
+            (if necessary); No subsets will be created from a datamanager but 
+            each train and validation set consists of a integer number of full 
+            datamanagers
+
+            * passing a single datamanager: the data within the single manager 
+            will be splitted and multiple datamanagers will be created holding
```
Comment (ORippler) [inline delira/training/experiment.py]: typo: will be split

File: delira/training/experiment.py [hunk 2]
```diff
@@ -94,17 +94,31 @@ class initialization
...
 
         Parameters
         ----------
         num_epochs : int
             number of epochs to train the model
-        data : list of BaseDataManager
-            list of datamanagers (will be split for crossvalidation)
+        data : list of :class:`BaseDataManager` or single 
+               :class:`BaseDataManager`
+            list of datamanagers or single datamanger
```
Comment (ORippler) [inline delira/training/experiment.py]: typo datamanger

File: delira/training/experiment.py [hunk 3]
```diff
@@ -118,20 +132,48 @@ def kfold(self, num_epochs: int, data: typing.List[BaseDataManager],
 
         """
 
+        # set number of splits if not specified
         if num_splits is None:
-            num_splits = len(data)
-
+            if isinstance(data, BaseDataManager):
+                num_splits = len(data.dataset)
```
Comment (ORippler) [inline delira/training/experiment.py]: I don't think setting num_splits = len(dataset) is a good idea. If for example we have 5000 training samples we will get 5000 splits. Better throw a warning and reduce it to some commonly used number like 5 or at most 10.

**Review by ORippler — CHANGES_REQUESTED** [review]
In general: How easy is it to extract the labels from `datamanger.dataset` ? If so, one could think of also adding `StratifiedKfold` to ensure equal label distributions amongst splits. I think with `BaseLazyDataset` this could probe to be difficult though.

**Comment by justusschock** [comment]
To add a `StratifiedKfold` we would have to. Iterate over all data samples once. While this could take a while, it should not be difficult to implement.

We could raise a Warning if the dataset is a lazy one, but this overhead must be made only once, I think we should add it.

File: delira/training/experiment.py (hunk 1 above)
Comment (justusschock) [reply delira/training/experiment.py]: Fixed

File: delira/training/experiment.py (hunk 2 above)
Comment (justusschock) [reply delira/training/experiment.py]: fixed

File: delira/training/experiment.py (hunk 3 above)
Comment (justusschock) [reply delira/training/experiment.py]: Changed default value to 10 splits

**Comment by justusschock** [comment]
cc @ORippler . Requested changes have been made

File: delira/training/experiment.py [hunk 4]
```diff
@@ -177,6 +179,74 @@ def kfold(self, num_epochs: int,
...
+        Parameters
+        ----------
+        num_epochs : int
+            number of epochs to train the model
+        data : :class:`BaseDataManager`
+            single datamanager
+            (will be split for crossvalidation)
+        num_splits : None or int
+            number of splits for kfold
+            if None: len(data) splits will be validated
```
Comment (ORippler) [inline delira/training/experiment.py]: if None: 10 splits

**Review by ORippler — CHANGES_REQUESTED** [review]
update docstring.

also [here](https://github.com/justusschock/delira/blob/e5dc2e79aa9281b5d9d931b62f5430f7ec01933a/delira/training/experiment.py#L124)

File: delira/training/experiment.py (hunk 4 above)
Comment (justusschock) [reply delira/training/experiment.py]: done

**Review by ORippler — APPROVED** [review]
Seems good to me now
